express/parsers/apps/aiida/formats/zip.py

```diff
@@ -0,0 +1,171 @@
+                self._metadata = json.loads(source.read('metadata.json'))
+        return self._metadata
+
+    @property
+    def data(self):
+        if self._data is None:
+            with ZipFile(self.zip_file_path) as source:
+                self._data = json.loads(source.read('data.json'))
+        return self._data
+
+    def _check_supported_versions(self):
+        """
+        Check whether the versions of the exported archive file are supported.
+
+        Args:
+            metadata (dict): deserialized metadata.json object
+
+        Raises:
+            RuntimeError
+        """
+        try:
+            aiida_version = self.metadata['aiida_version']
+            export_version = self.metadata['export_version']
+        except KeyError as error:
+            raise BadAiidaArchiveZipFile(self.zip_file_path, error)
+
+        if self.metadata['aiida_version'] != SUPPORTED_AIIDA_VERSION:
+            raise AiidaArchiveFileParseError(
+                path=self.zip_file_path,
+                reason=f"aiida_version ({metadata['aiida_version']}) not supported. "
+                f"Supported: {SUPPORTED_AIIDA_VERSION}")
+
+        if self.metadata['export_version'] != SUPPORTED_AIIDA_ARCHIVE_VERSION:
+            raise AiidaArchiveFileParseError(
+                path=self.zip_file_path,
+                rason=f"archive export_version ({metadata['export_version']}) not supported. "
+                f"Supported: {SUPPORTED_AIIDA_ARCHIVE_VERSION}")
+
+    def structures(self):
+        """
+        Extract all structures from compatible AiiDA archive zip files.
+
+        Returns:
+            list
+        """
+
+        return list(self._gather_structures(self.data))
+
+    @classmethod
+    def _gather_structures(cls, data):
+        """
+        Yield parsed structure objects from nodes within AiiDA archive.
+
+        Args:
+            data (dict): deserialized data.json object
+        Yields:
+            dict
+        """
+
+        nodes = data['export_data']['Node']
+        structure_nodes = {pk: node for (pk, node) in nodes.items()
+                           if node['node_type'] == 'data.structure.StructureData.'}
+
+        for pk in structure_nodes:
+            export_data = data['export_data']['Node'][pk]
+            node_attributes = data['node_attributes'][pk]
+            yield cls._parse_structure_node_attributes(export_data, node_attributes)
+
+    @staticmethod
+    def _parse_structure_node_attributes(export_data, attributes):
+        """
+        Parse an AiiDA StructureData node.
+        """
+
+        # Prepare mapping
+        assert export_data['node_type'] == 'data.structure.StructureData.'
+
+        sites = list(enumerate(attributes['sites']))
+        kinds = {kind['name']: kind for kind in attributes['kinds']}
+        assert all(len(kind['symbols']) == 1 for kind in kinds.values())
+
+        instance = json.loads(TEMPLATES_MATERIALS_PATH.read_text())
+        instance.update({
```

Comment:
What I meant is:

- inside the template below - `materials.json` (should be called `material.json` btw) we can add template variables, for example `{ ID }`, or `{ LATTICE_VECTOR_A }` like:
```
{
    "_id": "{{ ID }}"
    "lattice": {
       "vectors": {
        { "a": "{{ LATTICE_VECTOR_A }}" ...
   ...
}
```
- then, when we want to generate the instance from it we read text, render it and load:

```python
json.loads(Template.render(template, ID=export_data['uuid'], LATTICE_VECTORS_A=attributes['cell'][0]
...)

Reply:
I've applied that change in https://github.com/Exabyte-io/express/pull/56/commits/7aa402b0cc223ba75a9fa3dcbf35eea96499fb80 .